manager/allocator/network/port/port_test.go

```diff
@@ -0,0 +1,955 @@
+							},
+						},
+					}
+				})
+				It("should return ErrPortSpaceExhausted", func() {
+					_, err := pa.Allocate(endpoint, spec)
+					Expect(err).To(HaveOccurred())
+					Expect(err).To(WithTransform(
+						func(e error) bool { return IsErrPortSpaceExhausted(e) },
+						BeTrue(),
+					))
+					Expect(err.Error()).To(Equal("the dynamically allocated port space [30000,32767] is exhausted for protocol TCP"))
+				})
+			})
+		})
+
+		Context("when a port in the dynamic range desired to be statically allocated", func() {
+			var (
+				spec *api.EndpointSpec
+				p    Proposal
+				err  error
+			)
+			BeforeEach(func() {
+				// NOTE(dperny): this test relies on the fact that dynamic port
+				// allocation is sequential. If the behavior of dynamic port
+				// allocation changes, this test will need to be altered.
+				spec = &api.EndpointSpec{
+					Ports: []*api.PortConfig{
+						{
+							Protocol:    api.ProtocolTCP,
+							TargetPort:  80,
+							PublishMode: api.PublishModeIngress,
+						},
+						{
+							Protocol:      api.ProtocolTCP,
+							TargetPort:    80,
+							PublishedPort: DynamicPortStart,
+							PublishMode:   api.PublishModeIngress,
+						},
+					},
+				}
+				p, err = pa.Allocate(&api.Endpoint{}, spec)
+			})
+			It("should succeed", func() {
+				Expect(err).ToNot(HaveOccurred())
+				Expect(p).ToNot(BeNil())
+				Expect(p.IsNoop()).ToNot(BeTrue())
+				Expect(p.Ports()).ToNot(BeEmpty())
+			})
+			It("should give the statically allocated port first priority", func() {
+				Expect(p.Ports()[1].PublishedPort).To(Equal(DynamicPortStart))
+				Expect(p.Ports()[0].PublishedPort).To(And(
+					BeNumerically(">=", DynamicPortStart),
+					BeNumerically("<=", DynamicPortEnd),
+					Not(Equal(DynamicPortStart)),
+				))
+			})
+		})
+		Context("when changing ports", func() {
+			var (
+				endpoint *api.Endpoint
+				spec     *api.EndpointSpec
+			)
+			BeforeEach(func() {
+				spec = &api.EndpointSpec{
+					Ports: []*api.PortConfig{
+						{
+							Protocol:    api.ProtocolTCP,
+							TargetPort:  80,
+							PublishMode: api.PublishModeIngress,
+						},
+						{
+							Protocol:      api.ProtocolTCP,
+							TargetPort:    80,
+							PublishedPort: 80,
+							PublishMode:   api.PublishModeIngress,
+						},
+					},
+				}
+				endpoint = &api.Endpoint{
+					Spec:  spec,
+					Ports: spec.Ports,
+				}
+				endpoint.Ports[0].PublishedPort = DynamicPortStart
+				pa.Restore([]*api.Endpoint{endpoint})
+			})
+			Context("from dynamic to statically allocated", func() {
+				var (
+					specCopy *api.EndpointSpec
+					p        Proposal
+					e        error
+				)
+				BeforeEach(func() {
+					specCopy = spec.Copy()
+					specCopy.Ports[0].PublishedPort = 8080
+
+					p, e = pa.Allocate(endpoint, specCopy)
+				})
+				It("should succeed", func() {
+					Expect(e).ToNot(HaveOccurred())
+					Expect(p).ToNot(BeNil())
+					Expect(p.IsNoop()).ToNot(BeTrue())
+					Expect(p.Ports()).ToNot(BeEmpty())
```

Comment:
Should we add more assertions that the previous dynamic allocation was freed?

Reply:
Added this assertion.